### ISSUE TITLE
Consider additional field in signatures noting encoding that signature is over (e.g. der or json)

### ISSUE THREAD

**Issue by awwad** [uptane/obsolete-reference-implementation#54]
Previous to adding ASN.1/DER encoding, signable objects (general form :`{'signed': ..., 'signatures': []}`) in TUF and Uptane would always contain signatures over a canonical JSON encoding of the value of the 'signed' element.

When ASN.1/DER is used, the signature is instead over the data encoded into DER after having been converted into a pyasn1 representation matching the ASN.1 definitions (captured in uptane/encoding/asn1_definitions.py and tuf/encoding/asn1_definitions.py), for important reasons. Despite this, the 'signed' element of the dictionary in memory still contains data in the (old, comprehensible, clear) format specified by formats.py, so that metadata can be used. Validating signatures requires that the encoding used to sign over be known, and this leads to a possibly confusing situation.

I'm considering adding to individual signatures a field that identifies the way the data was encoded when it was signed.

```javascript
{
   'signed': ...,
   'signatures': [
      {
         'keyid': '1234...',
         'method': 'ed25519',
         'sig': 'abcdef0...'},
         'sig_over_encoding': 'der'   // Proposed new optional field
      {...},
   ]
}
```

**Comment by awwad** [comment]
### Example:
Consider the case of a Vehicle Manifest signed over DER and provided to the Director server and so added to the InventoryDB. The VVM will show up in Uptane's typical Python dictionary format, with a 'signed' and 'signatures' field, with the 'signed' field in the friendly format of uptane.formats.VEHICLE_VERSION_MANIFEST_SCHEMA. There is no obvious way to know that the signature is over the (hash of the) DER encoding of the 'signed' data. While the system is running in DER mode, it will assume that. If it later runs in JSON mode, the validation will instead fail.

If we label the signature indicating it's over DER, the Director will always know what to do at a later date.

**Comment by awwad** [comment]
### Alternatives:
1. Director checks the signature with both encodings and is happy with it if either works
2. Director could default to whatever mode it's operating in and try the other if that fails.
3. We could expand the 'method' field with suffixes or something ugly like that (`'method': 'ed25519 over DER'`)

**Comment by JustinCappos** [comment]
In general, the OEM is going to have defined this.  So for the demo
implementation, it shouldn't matter what you decide to do.

I wonder if this is more of a TUF problem (if it is an issue there), than
an Uptane one?

**Comment by awwad** [comment]
In the Uptane demo implementation, I've gone with alternative 0: I simply check employing the configured method, expecting the whole system to use all DER or all JSON.

Yes, I think this is more relevant for TUF if TUF chooses to support multiple encodings. It would require a TAP. (Though I might just insert it in the TAP for the first encoding we add, which I suppose will end up being ASN.1/DER.)

**Comment by awwad** [comment]
(BTW, if an OEM is going to support signing over multiple slightly different encodings - which they may well find it beneficial to do with disparate components - then for their own purposes, they may end up having such a field in each signature.

But while I suppose it would be nice to demonstrate this in the code to guide them, it's still more complexity with its own tradeoff. On the first hand, another advantage might be that it makes the nature of the signatures (over encodings) a bit more noticeable in the code, which might be good for other implementers. I know I've managed to confuse myself by forgetting about encoding details before.)

**Comment by JustinCappos** [comment]
Right, but I think we need not require them to do so.  Hence I think
whatever you do in your demo is fine, but it wouldn't be in the spec.
